magento/magento2
# keeps on product in customer or visitor compare list if stock status is changing to out_of_stock (related to issues 24620 and 24617 + fix to 24582)

### Description
Related to issues #24620 and #24617
This fix provide changing logic of Product Compare functionality. It's alows to keeps on  product visibility that already in Compare List when it Stock Status changing to OUT_OF_STOCK
**upd:** Added fix related to issue magento/magento2#24582   
**upd:** Added filter to exclude disabled items from collection

### Manual testing scenarios (*)
1. Find the product **Chaz Kangeroo Hoodie** (SKU: MH01) from the Storefront and add to compare
2. Login to Admin
3. Find the product in the admin panel by SKU: MH01 and open it
4. Expand Configurations tab
5. Click **Edit Configurations** > **Next** > **Next**
6. Under **Quantity** enable **Apply single quantity to each SKUs** and set it to 0
7. Click **Next** > **Generate Products**
8. Save the product.
9. Copy the URL key of the product and construct the products URL:
{{base_url}} + / + {{URL_key}} + {{product_URL_suffix}}, for example http://graphql-ce.loc/chaz-kangeroo-hoodie.html
Verify that it has OUT_OF_STOCK status.
10. Navigate to {{base_url}}/catalog/product_compare/

### Actual result (*)

1.  We see in the top right that there is one item in the comparison list  
![Screenshot_20190916_140835](https://user-images.githubusercontent.com/11827230/64958488-f5fa6580-d897-11e9-816f-51f858b03090.png)

### Contribution checklist (*)
